### DISCUSSION TITLE
creating-a-source-generator-part-2-testing-an-incremental-generator-with-snapshot-testing/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#39] [Announcements]
# creating-a-source-generator-part-2-testing-an-incremental-generator-with-snapshot-testing/

In this post I describe how to unit test an incremental generator using Verify for snapshot testing

https://andrewlock.net/creating-a-source-generator-part-2-testing-an-incremental-generator-with-snapshot-testing/

**Comment by goldytech** [comment]
Hi Andrew,
Thanks for this awesome content. 
When debugging the unit tests of my Generator, the first transformation `IsSyntaxTargetForGeneration(SyntaxNode syntaxNode)` For me the syntax mode is of type `CompilationUnitSyntax` and not the `ClassDeclarationSyntax` . Hence the return value is False and the transformation fails. Any idea how this can be resolved.

**Comment by StarkeContrast** [comment]
Getting a failure on that same line, not 'NotAnAttributeType', but literally an empty collection.  It cannot find the marker Attribute.  The syntax tree for the 'context' variable is just the enum code from the test, and it absolutely cannot find the marker Attribute.  Thoughts?

**Reply by StarkeContrast** [reply]
It looks like the SemanticModel property is scoped to the current node (the Enum type) not the global compilation scope.  To go searching through the global compilation, there's a Compilation property hanging off the SemanticModel, but the APIs for searching for symbols differs slightly:

var match = context.SemanticModel.Compilation.GetSymbolsWithName(name =>
	name == $"{attributeSyntax.ToFullString()}{attributeSyntax.Kind()}")
	.SingleOrDefault();
if(match is not INamedTypeSymbol attributeContainingTypeSymbol)
{
	// weird, we couldn't get the symbol, ignore it
	continue;
}

//INamedTypeSymbol attributeContainingTypeSymbol = attributeSymbol.ContainingType;
string fullName = attributeContainingTypeSymbol.ToDisplayString();

[a comment by an automated account is not shown]

**Comment by KeithLRobertson** [comment]
This is the best thing since sliced cheese! Was having a heckuva time building an incremental generator. "Verify" makes it much easier, where I can set up tests to step through and look at semantic objects in the debugger. Many thanks for explaining how to set up and use this awesome tool!

**Comment by aateeque** [comment]
Really useful & detailed series; very helpful thanks. 
Just needs a bit more highlighting about the assembly references part when writing custom Generators; I found myself spending a lot of time with that; @stuart Whitefords comments below were helpful; but came across this really helpful method https://github.com/DavidFineboym/LoggingDecoratorGenerator/blob/18c2888d9f5e87345871cfed11147bfdf7a32613/LoggingDecoratorGenerator.Tests/TestHelper.cs#L14 which solves all assembly reference problems

**Comment by alexbkornev** [comment]
We develop generators for primitive code like boilerplate constructors and prefer maintaining their test baselines directly in C# code rather than individual snapshot files. Then we can rewrite them manually or with Scand Storm Petrel. Can Verify rewrite baselines similar way? Storm Petrel is pretty new and might be risky tool what might switch us to manual baseline changes in the future.
